From 59fa69d26f940d6931d4fd7dce747f5095f11441 Mon Sep 17 00:00:00 2001
From: Patrick Yavitz <pyavitz@gmail.com>
Date: Sat, 9 Aug 2025 07:00:05 -0400
Subject: [PATCH] Add syslinux and script support

The board/spacemit/k1-x/k1-x.env file is deleted by the builder
before the patch is applied.

Signed-off-by: Patrick Yavitz <pyavitz@gmail.com>
---
 board/spacemit/k1-x/k1-x.env | 26 ++++++++++++++++++++++++++
 board/spacemit/k1-x/k1x.c    | 20 ++++++++++----------
 configs/k1_defconfig         |  1 +
 include/configs/k1-x.h       | 24 ++++++++++--------------
 4 files changed, 47 insertions(+), 24 deletions(-)
 create mode 100644 board/spacemit/k1-x/k1-x.env

diff --git a/board/spacemit/k1-x/k1-x.env b/board/spacemit/k1-x/k1-x.env
new file mode 100644
index 00000000..67c5f1b7
--- /dev/null
+++ b/board/spacemit/k1-x/k1-x.env
@@ -0,0 +1,26 @@
+// SYSLINUX and SCRIPT Support
+autoboot=if test -e ${devtype} ${devnum}:${distro_bootpart} /extlinux/extlinux.conf; then \
+		sysboot ${devtype} ${devnum}:${distro_bootpart} any 0x2000000 /extlinux/extlinux.conf; \
+	elif test -e ${devtype} ${devnum}:${distro_bootpart} /boot/extlinux/extlinux.conf; then \
+		sysboot ${devtype} ${devnum}:${distro_bootpart} any 0x2000000 /boot/extlinux/extlinux.conf; \
+	elif test -e ${devtype} ${devnum}:${distro_bootpart} /boot.scr; then \
+		load ${devtype} ${devnum}:${distro_bootpart} 0x2000000 /boot.scr; source 0x2000000; \
+	elif test -e ${devtype} ${devnum}:${distro_bootpart} /boot/boot.scr; then \
+		load ${devtype} ${devnum}:${distro_bootpart} 0x2000000 /boot/boot.scr; source 0x2000000; \
+	fi;
+	nvme part; \
+	if test -e nvme 0:1 /extlinux/extlinux.conf; then \
+		sysboot nvme 0:1 any 0x2000000 /extlinux/extlinux.conf; \
+	elif test -e nvme 0:1 /boot/extlinux/extlinux.conf; then \
+		sysboot nvme 0:1 any 0x2000000 /boot/extlinux/extlinux.conf; \
+	elif test -e nvme 0:1 /boot.scr; then \
+		load nvme 0:1 0x2000000 /boot.scr; source 0x2000000; \
+	elif test -e nvme 0:1 /boot/boot.scr; then \
+		load nvme 0:1 0x2000000 /boot/boot.scr; source 0x2000000; \
+	fi;
+
+bootcmd=echo "Loading K1-X Environment ..."; \
+	echo ""; \
+	echo "Running NVMe Scan ..."; \
+	nvme scan; \
+	run autoboot
diff --git a/board/spacemit/k1-x/k1x.c b/board/spacemit/k1-x/k1x.c
index b2fc28c8..4c788689 100644
--- a/board/spacemit/k1-x/k1x.c
+++ b/board/spacemit/k1-x/k1x.c
@@ -530,8 +530,8 @@ void _load_env_from_blk(struct blk_desc *dev_desc, const char *dev_name, int dev
 #endif
 	}
 
-	env_set("bootfs_part", simple_itoa(part));
-	env_set("bootfs_devname", dev_name);
+	env_set("distro_bootpart", simple_itoa(part));
+	env_set("devnum", dev_name);
 
 	/*load env.txt and import to uboot*/
 	memset((void *)CONFIG_SPL_LOAD_FIT_ADDRESS, 0, CONFIG_ENV_SIZE);
@@ -734,7 +734,7 @@ void setenv_boot_mode(void)
 	u32 boot_mode = get_boot_mode();
 	switch (boot_mode) {
 	case BOOT_MODE_NAND:
-		env_set("boot_device", "nand");
+		env_set("devtype", "nand");
 		break;
 	case BOOT_MODE_NOR:
 		char *blk_name;
@@ -745,23 +745,23 @@ void setenv_boot_mode(void)
 			return;
 		}
 
-		env_set("boot_device", "nor");
-		env_set("boot_devnum", simple_itoa(blk_index));
+		env_set("devtype", "nor");
+		env_set("devnum", simple_itoa(blk_index));
 		break;
 	case BOOT_MODE_EMMC:
-		env_set("boot_device", "mmc");
-		env_set("boot_devnum", simple_itoa(MMC_DEV_EMMC));
+		env_set("devtype", "mmc");
+		env_set("devnum", simple_itoa(MMC_DEV_EMMC));
 		break;
 	case BOOT_MODE_SD:
-		env_set("boot_device", "mmc");
-		env_set("boot_devnum", simple_itoa(MMC_DEV_SD));
+		env_set("devtype", "mmc");
+		env_set("devnum", simple_itoa(MMC_DEV_SD));
 		break;
 	case BOOT_MODE_USB:
 		// for fastboot image download and run test
 		env_set("bootcmd", CONFIG_BOOTCOMMAND);
 		break;
 	default:
-		env_set("boot_device", "");
+		env_set("devtype", "");
 		break;
 	}
 }
diff --git a/configs/k1_defconfig b/configs/k1_defconfig
index 4f419bc1..05337aaf 100644
--- a/configs/k1_defconfig
+++ b/configs/k1_defconfig
@@ -291,3 +291,4 @@ CONFIG_PRINT_TIMESTAMP=y
 # CONFIG_SPL_SHA256 is not set
 CONFIG_ZSTD=y
 # CONFIG_HEXDUMP is not set
+CONFIG_OF_LIBFDT_OVERLAY=y
diff --git a/include/configs/k1-x.h b/include/configs/k1-x.h
index b15d2e0b..edad0e55 100644
--- a/include/configs/k1-x.h
+++ b/include/configs/k1-x.h
@@ -106,11 +106,6 @@
 #define TLV_CODE_EEPROM_I2C_INDEX	0x81
 #define TLV_CODE_EEPROM_PIN_GROUP	0x82
 
-// #define RAMDISK_LOAD_ADDR		(CONFIG_FASTBOOT_BUF_ADDR + CONFIG_FASTBOOT_BUF_SIZE)
-// #define DTB_LOAD_ADDR		(CONFIG_FASTBOOT_BUF_ADDR + CONFIG_FASTBOOT_BUF_SIZE * 2)
-#define RAMDISK_LOAD_ADDR		0x21000000
-#define DTB_LOAD_ADDR			0x31000000
-
 // for those has NOT been through test procedure(ATE)
 #define SVT_DRO_DEFAULT_VALUE	(120)
 
@@ -173,6 +168,11 @@ enum private_part_offset {
 #define BOOT_TARGET_DEVICES(func) \
 	func(QEMU, qemu, na)
 
+#define KERNEL_ADDR_R		__stringify(0x10000000)
+#define RAMDISK_ADDR_R		__stringify(0x21000000)
+#define FDT_ADDR_R		__stringify(0x31000000)
+#define FDTOVERLAY_ADDR_R	__stringify(0x01000000)
+
 #include <config_distro_bootcmd.h>
 
 #define BOOTENV_DEV_QEMU(devtypeu, devtypel, instance) \
@@ -198,17 +198,13 @@ enum private_part_offset {
 /*if env not use for spl, please define to board/spacemit/k1-x/k1-x.env */
 #define CONFIG_EXTRA_ENV_SETTINGS \
 	"stdout_flash=serial,vidconsole\0" \
-	"kernel_comp_addr_r=" __stringify(CONFIG_SYS_LOAD_ADDR) "\0" \
-	"kernel_comp_size=" __stringify(CONFIG_FASTBOOT_BUF_SIZE) "\0" \
-	"kernel_addr_r=" __stringify(CONFIG_FASTBOOT_BUF_ADDR) "\0" \
-	"ramdisk_addr=" __stringify(RAMDISK_LOAD_ADDR) "\0" \
-	"dtb_addr=" __stringify(DTB_LOAD_ADDR) "\0" \
 	"scriptaddr=0x2c100000\0" \
 	"pxefile_addr_r=0x0c200000\0" \
-	"splashimage=" __stringify(CONFIG_FASTBOOT_BUF_ADDR) "\0" \
-	"splashpos=m,m\0" \
-	"splashfile=bianbu.bmp\0" \
-	"led0_gpio=" __stringify(STATUS_LED_GPIO0) "\0" \
+	"fdt_addr_r=" FDT_ADDR_R "\0" \
+	"kernel_addr_r=" KERNEL_ADDR_R "\0" \
+	"ramdisk_addr_r=" RAMDISK_ADDR_R "\0" \
+	"fdtoverlay_addr_r=" FDTOVERLAY_ADDR_R "\0" \
+	"fdtfile=spacemit/" CONFIG_DEFAULT_DEVICE_TREE ".dtb\0" \
 	BOOTENV_DEVICE_CONFIG
 
 
-- 
2.43.0

